pkg/realtime/memhub.go

```diff
@@ -0,0 +1,95 @@
+package realtime
+
+import "sync"
+
+type hub struct {
+	sync.Mutex
+
+	// topics should only be manipulated by the hub loop
+	// it is a Map(instance + type -> Topic)
+	instanceTopics map[string]*topic
+
+	// Map(domain -> Topic)
+	globalTopics map[string]*topic
+}
+
+func (h *hub) Publish(e *Event) {
+	panic("Wrong usage : you should not publish on central hub.")
+}
+
+func (h *hub) Subscribe(t string) EventChannel {
+	h.Lock()
+	defer h.Unlock()
+	gt, ok := h.globalTopics[t]
+	if !ok {
+		h.globalTopics[t] = makeTopic(h, t)
+		gt = h.globalTopics[t]
+	}
+	sub := makeSub([]*topic{gt})
+	go func() { gt.subscribe <- sub }()
+
+	return sub
+}
+
+func (h *hub) remove(t *topic) {
+	h.Lock()
+	defer h.Unlock()
+	delete(h.globalTopics, t.key)
+	delete(h.instanceTopics, t.key)
+}
+
+type instancehub struct {
+	instance string
+	mainHub  *hub
+}
+
+func (ih *instancehub) Publish(e *Event) {
+	ih.mainHub.Lock()
+	defer ih.mainHub.Unlock()
```

Comment:
I would add `getInstanceTopics/getGlobalTopics` on the `hub` struct and handle the mutex (un)locking from there. Also it looks like we could maybe benefit from a `RWMutex` (since Publish is read-only) and maybe one mutex per map.